### PR TITLE
Add types for shortcode package

### PR DESCRIPTION
**Pull request by manzoorwanijk**
## What?
Add TS types for shortcode package.

## Why?
To have better type safety and to unblock `blocks` package being converted to TS

## How?
Unlike #60526, this PR extracts the JSDoc types to a `types.ts` file to expose types without making changes to the actual JS code.

## Testing Instructions
- Verify that CI is happy

## Next steps

- Deprecate [`@types/wordpress__shortcode`](https://www.npmjs.com/package/@types/wordpress__shortcode) after this PR gets merged and released to npm.
- Convert `shortcode` `Object.assign` to JS Class declaration for better code quality.

### CODE REVIEW

**Comment by gziolo** [comment]
> However, the next steps should be to convert shortcode Object.assign to JS Class declaration for better code quality.

Sounds fine, but only makes sense if the file gets converted fully to TypeScript.

File: packages/shortcode/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+export type Match =
+	| NonNullable< ReturnType< RegExp[ 'exec' ] > >
+	| Array< string >;
+
+export type ReplaceCallback = ( shortcode: Shortcode ) => string;
+
+/**
+ * WordPress Shortcode instance.
+ */
+export interface WPShortcode {
```
Comment (gziolo) [inline packages/shortcode/src/types.ts]: Can we drop somehow this WP prefix? It doesn’t add any value in TypeScript as these names aren’t global like in JSDoc. It also differs on DefinitelyTyped:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/c39e00d70995131a814a23b8054ea191d910925b/types/wordpress__shortcode/index.d.ts#L116

**Comment by manzoorwanijk** [comment]
> Sounds fine, but only makes sense if the file gets converted fully to TypeScript.

Yes, that's the plan.

**Review by gziolo — APPROVED** [review]
Nice one. Thank you for working on it.

I had only one question. We also should update the changelog to include the note that types are now available. 🎉

File: packages/shortcode/src/types.ts (hunk 1 above)
Comment (manzoorwanijk) [reply packages/shortcode/src/types.ts]: `Shortcode` is already used above, that's why I used that. We can get rid of it once we convert that file to TS.

Comment (gziolo) [reply packages/shortcode/src/types.ts]: We only need to do it in one release cycle so folks don't have to change these types after the get published to npm 👍

Comment (manzoorwanijk) [reply packages/shortcode/src/types.ts]: Alright, I have removed the `WP` prefix. Thanks

**Comment by manzoorwanijk** [comment]
Adding changelog now

**Review by gziolo — APPROVED** [review]
Awesome, thank you!

**Pull request merged**